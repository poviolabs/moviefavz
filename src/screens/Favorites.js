import React from 'react';

import { observer } from 'mobx-react';
import { useStores } from '../hooks';
import { useAuth0 } from '@auth0/auth0-react';

import { saveAs } from 'file-saver';

import { Layout, Typography, Row, Col, Button } from 'antd';
import { ExportOutlined } from '@ant-design/icons';

import { Container, Section } from '../components/layout';
import { Emoji } from '../components/typography';
import { MoviesGrid, LoginButton, HomeButton } from '../components/ui';
import { STATE_TYPES } from '../constants';

const { Content } = Layout;
const { Title, Paragraph } = Typography;

const Favorites = () => {
  const { moviesStore } = useStores();
  const { isAuthenticated, isLoading, user } = useAuth0();

  const hasFavorites = React.useMemo(() => {
    return moviesStore.favoritesPreviews.length > 0;
  }, [moviesStore.favoritesPreviews]);

  const stateLoading = React.useMemo(() => {
    return moviesStore.state === STATE_TYPES.pending || isLoading;
  }, [moviesStore.state, isLoading]);

  const handleFavzExport = () => {
    const data = moviesStore.favorites.reduce(
      (arr, id) => [...arr, moviesStore.singleMoviesById[id]],
      []
    );

    const fileName = `${
      user.given_name ? user.given_name : user.nickname
    }_Movie_Favz.json`;

    const fileToSave = new Blob([JSON.stringify(data, null, 2)], {
      type: 'application/json',
    });

    saveAs(fileToSave, fileName);
  };

  return (
    <Content>
      <Container>
        <Section>
          <Row align="middle">
            <Col flex="auto">
              <Title level={1}>Your Favz</Title>
              {hasFavorites || stateLoading ? (
                <Paragraph>
                  Here are your favorited movies. It&apos;s always smart to keep
                  track of things you like{' '}
                  <Emoji emoji="👇" label="pointing finger" />
                </Paragraph>
              ) : isAuthenticated ? (
                <>
                  <Paragraph>
                    Looks like you don&apos;t have any favorited movies just yet{' '}
                    <Emoji emoji="🤷‍♂️" label="man shrugging" />. Return to the
                    app home and find some movies you like.
                  </Paragraph>
                  {moviesStore.state === STATE_TYPES.done && <HomeButton />}
                </>
              ) : (
                <>
                  <Paragraph>
                    In order to add movies to favorites and view your recent
                    searches, you need to authenticate. Click the Login button
                    bellow, to start saving your movies.
                  </Paragraph>
                  <LoginButton />
                </>
              )}
            </Col>
            <Col flex="200px">
              {isAuthenticated && hasFavorites && (
                <Button
                  block
                  size="large"
                  type="primary"
                  onClick={handleFavzExport}
                >
                  Export your Favz <ExportOutlined />
                </Button>
              )}
            </Col>
          </Row>
        </Section>
        {hasFavorites && <MoviesGrid movies={moviesStore.favoritesPreviews} />}
      </Container>
    </Content>
  );
};

export default observer(Favorites);
