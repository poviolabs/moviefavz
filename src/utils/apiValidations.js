const validResponse = (status) => status && status >= 200 && status < 300;

export default { validResponse };
