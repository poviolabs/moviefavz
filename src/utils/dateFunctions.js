import dayjs from 'dayjs';

const getUnixDay = (timestamp = Date.now()) => {
  return dayjs(timestamp).startOf('date').valueOf();
};

const fillDaysToMatch = (num, data) => {
  if (!data.length) {
    return [];
  }
  const [lastDate] = data.reverse();

  return Array.from('x'.repeat(num - data.length))
    .map((_, i) => i + 1)
    .reverse()
    .map((i) => {
      return {
        value: 0,
        timestamp: dayjs(lastDate.timestamp).subtract(i, 'day').valueOf(),
      };
    });
};

export default { getUnixDay, fillDaysToMatch };
