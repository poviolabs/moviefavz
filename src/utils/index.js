export { default as arrayFunction } from './arrayFunctions';
export { default as apiValidations } from './apiValidations';
