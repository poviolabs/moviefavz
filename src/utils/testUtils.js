import React from 'react';
import { render } from '@testing-library/react';

import { BrowserRouter } from 'react-router-dom';

import { ThemeProvider } from 'styled-components';
import { theme } from '../styles';

// eslint-disable-next-line react/prop-types
const AllTheProviders = ({ children }) => {
  return (
    <ThemeProvider theme={theme}>
      <BrowserRouter>{children}</BrowserRouter>
    </ThemeProvider>
  );
};

const customRender = (ui, options) =>
  render(ui, { wrapper: AllTheProviders, ...options });

// re-export everything
export * from '@testing-library/react';

// override render method
export { customRender as render };
