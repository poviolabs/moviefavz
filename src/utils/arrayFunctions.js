import dateFunctions from './dateFunctions';

const groupArr = (data, size) => {
  const group = [];
  for (let i = 0, j = 0; i < data.length; i++) {
    if (i >= size && i % size === 0) {
      j++;
    }
    group[j] = group[j] || [];
    group[j].push(data[i]);
  }
  return group;
};

const groupByDay = (data, key = 'timestamp') => {
  const groupedByKey = data.reduce((obj, item) => {
    const timestamp = item[key];
    const day = dateFunctions.getUnixDay(timestamp);

    if (Object.prototype.hasOwnProperty.call(obj, day)) {
      obj[day].value += 1;
    } else {
      obj[day] = {
        value: 1,
        timestamp,
      };
    }
    return obj;
  }, {});
  return Object.values(groupedByKey);
};

export default { groupArr, groupByDay };
