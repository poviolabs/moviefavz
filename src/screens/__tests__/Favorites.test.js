import React from 'react';
import { render } from '../../utils/testUtils';

import Favorites from '../Favorites';

describe('<Favorites />', () => {
  it('Renders the Component', () => {
    const props = {};
    const container = render(<Favorites {...props} />);
    expect(container.firstChild).toMatchSnapshot();
  });
});
