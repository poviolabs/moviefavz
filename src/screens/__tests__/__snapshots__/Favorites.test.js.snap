// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Favorites /> Renders the Component 1`] = `undefined`;
